sled-agent/src/bootstrap/rack_secret.rs

```diff
@@ -0,0 +1,171 @@
+/// and that there is an automatic mechanism to retrieve the decryption key.
+/// Furthermore, we must guarantee that the the key retrieval mechanism is not
+/// available when a only a subset of sleds or disks from a rack are available
+/// to an attacker. The mechanism we use to provide these guarantees is based on
+/// <https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing>. A threshold secret
+/// is generated of which individual disk encryption keys are derived, and each
+/// of `N` server sleds receives one share of the secret. `K` of these shares
+/// must be combinded in order to reconstruct the shared secret such that disks
+/// may be decrypted. If fewer than `K` shares are available, no information
+/// about the secret may be recovered, and the disks cannot be decrypted. We
+/// call the threshold secret the `rack secret`.
+///
+/// Inside a rack then, the sleds cooperate over secure channels in order to
+/// retrieve key shares and reconstruct the `rack secret`, the resulting derived
+/// encryption keys, and unlock their own local storage. We call this procedure
+/// `rack unlock`. The establishment of secure channels and the ability to trust
+/// the validity of a participating peer is outside the scope of this particular
+/// type and orthogonal to its implementation.
+pub struct RackSecret {
+    secret: NonZeroScalar,
+}
+
+// This is a secret. Let's not print it outside of tests.
+#[cfg(test)]
+impl Debug for RackSecret {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.secret.as_ref().fmt(f)
+    }
+}
+
+impl PartialEq for RackSecret {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_ref() == other.as_ref()
+    }
+}
+
+impl Eq for RackSecret {}
+
+/// A verifier used to ensure the validity of a given key share for an unknown
+/// secret.
+///
+/// We use verifiable secret sharing to detect invalid shares from being
+/// combined and generating an incorrect secret. Each share must be verified
+/// before the secret is reconstructed.
+//
+// This is just a wrapper around a FeldmanVerifier from the vsss-rs crate.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Verifier {
+    verifier: FeldmanVerifier<Scalar, ProjectivePoint>,
+}
+
+impl Verifier {
+    pub fn verify(&self, share: &Share) -> bool {
+        self.verifier.verify(share)
+    }
+}
+
+impl RackSecret {
+    /// Create a secret based on the NIST P-256 curve
+    pub fn new() -> RackSecret {
+        let mut rng = OsRng::default();
+        let sk = SecretKey::random(&mut rng);
+        RackSecret { secret: sk.to_secret_scalar() }
+    }
+
+    /// Split a secert into `total_shares` number of shares, where combining
+    /// `threshold` of the shares can be used to recover the secret.
+    pub fn split(
+        &self,
+        threshold: u8,
+        total_shares: u8,
+    ) -> Result<(Vec<Share>, Verifier), vsss_rs::Error> {
+        assert!(total_shares >= threshold);
```

Comment:
Here and below in `combine_shares`, it seems a little quirky to have this assertion rather than return an error, especially as this is undocumented behavior which could cause these functions to panic.

vsss_rs::Error has a `SharingLimitLessThanThreshold` variant - you wouldn't even need to create a new variant type. TBH, it would probably be returning this error for you already, if not for the assertion.

Reply:
Additionally, this would make it easier for us to explicitly test this case.

Reply:
I was going with the practice that this is definitely a programmer error and so we should fail as fast and as loudly as possible. There's really nothing you can do with a returned error than rewrite your code. 

I can certainly change it. Thank you for pointing out that the proper error already exists.